org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexer.java

```diff
@@ -196,7 +203,7 @@ private void purgeMethodStatements(TypeDeclaration type) {
 		AbstractMethodDeclaration[] methods = type.methods;
 		for (int j = 0, length = methods == null ? 0 : methods.length; j < length; j++) {
 			AbstractMethodDeclaration method = methods[j];
-			if (method != null && (method.bits & ASTNode.HasFunctionalInterfaceTypes) == 0) {
+			if (method != null && (method.bits & ASTNode.HasFunctionalInterfaceTypes) == 0 && !definesAnonymousType(method)) {
```

Comment:
I don't have the sources checked out locally and github code navigation doesn't help me with it: Can we use the same tag-bit HasFunctionalInterfaceTypes (maybe rename to HasFunctionalInterfaceTypesOrLocalDeclarations)  and make sure it is correctly assigned when a local class declaration is encountered?

Does the bug manifest itself with a local named class declaration, too?
```
	public static void main(String[] args) {
		class R implements Runnable {
			@Override
			public void run() {
				// here is the usage of the method
				Stream.empty().filter(MissingReference2::missingReference1).count();
			}
		};
	}
```
Does it make sense to have test cases that are deeper nested, too?

Reply:
> I don't have the sources checked out locally and github code navigation doesn't help me with it: Can we use the same tag-bit HasFunctionalInterfaceTypes (maybe rename to HasFunctionalInterfaceTypesOrLocalDeclarations) and make sure it is correctly assigned when a local class declaration is encountered?

I can check. If you can point me to the code I should be checking, that would speed things up.

> Does the bug manifest itself with a local named class declaration, too?
> 
> ```
> 	public static void main(String[] args) {
> 		class R implements Runnable {
> 			@Override
> 			public void run() {
> 				// here is the usage of the method
> 				Stream.empty().filter(MissingReference2::missingReference1).count();
> 			}
> 		};
> 	}
> ```

Yes, there is a snippet for this in the new test. The name of the method needs work, yes.

> Does it make sense to have test cases that are deeper nested, too?

There is one snippet for this in the new test.

Reply:
> There is one snippet for this in the new test.

Sorry, I missed that when reading through this.

> If you can point me to the code I should be checking, that would speed things up.

I'd love to but I cannot seem to find the piece where this tagbit is set on the ASTNode via the browser. I don't have a local JDT clone at hand right now.

Reply:
> I'd love to but I cannot seem to find the piece where this tagbit is set on the ASTNode via the browser. I don't have a local JDT clone at hand right now.

OK, I'll check where the flag is set tomorrow.

Reply:
Given this snippet:

```
	public static void main(String[] args) {
		class R implements Runnable {
			public void run() {
				// here is the usage of the method
				java.util.stream.Stream.empty().filter(TestGh432::missingReference).count();
			}
		}
		new R().run();
	}
```

The stack trace at which the flag `ASTNode.HasFunctionalInterfaceTypes` is set is:

```
"Java indexing" #60 daemon prio=5 os_prio=0 cpu=315.32ms elapsed=806.03s tid=0x00007ffff0c57000 nid=0xf269 at breakpoint [0x00007fff5bcfb000]
   java.lang.Thread.State: RUNNABLE
        at org.eclipse.jdt.internal.compiler.parser.Parser.markEnclosingMemberWithLocalOrFunctionalType(Parser.java:12767)
        at org.eclipse.jdt.internal.compiler.parser.Parser.consumeReferenceExpression(Parser.java:9362)
        at org.eclipse.jdt.internal.compiler.parser.Parser.consumeReferenceExpressionTypeForm(Parser.java:9300)
        at org.eclipse.jdt.internal.compiler.parser.Parser.consumeRule(Parser.java:7915)
        at org.eclipse.jdt.internal.compiler.parser.Parser.parse(Parser.java:13149)
        at org.eclipse.jdt.internal.compiler.parser.Parser.parse(Parser.java:13505)
        at org.eclipse.jdt.internal.compiler.ast.MethodDeclaration.parseStatements(MethodDeclaration.java:251)
        at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.parseMethods(TypeDeclaration.java:1143)
        at org.eclipse.jdt.internal.compiler.parser.Parser.getMethodBodies(Parser.java:12167)
        at org.eclipse.jdt.internal.compiler.SourceElementParser.parseCompilationUnit(SourceElementParser.java:1138)
        at org.eclipse.jdt.internal.core.search.indexing.SourceIndexer.indexDocument(SourceIndexer.java:122)
        at org.eclipse.jdt.internal.core.search.JavaSearchParticipant.indexDocument(JavaSearchParticipant.java:83)
        at org.eclipse.jdt.internal.core.search.indexing.IndexManager.indexDocument(IndexManager.java:664)
        at org.eclipse.jdt.internal.core.search.indexing.IndexManager$2.execute(IndexManager.java:1298)
        at org.eclipse.jdt.internal.core.search.processing.JobManager.run(JobManager.java:478)
        at java.lang.Thread.run(java.base@11.0.10/Thread.java:834)
```

So far I don't see how to traverse the AST upward. The stack `Parser.astStack` contains only the inner type and its method, not the outer type method that defines it:

```
class R implements Runnable {
}
```
```
public void run() {
}
```

The method that is being parsed (see the stack trace, `MethodDeclaration.java:251`) is `main`, i.e. we are at the correct spot. Maybe we can do something after that line...